### PR TITLE
[WIP][SPARK-14629][SPARK-14630][Build] Add support for custom scala style rules & Add rule PublicAbstractMethodsHaveTypeChecker

### PR DESCRIPTION
**Pull request by lw-lin**
## What changes were proposed in this pull request?
- [ ] automatically build and install custom rules
- [x] added a custom rule `PublicAbstractMethodsHaveTypeChecker` implementation, which would ensure public abstract methods have explicit return types
## How was this patch tested?

the list of changes-to-make found by this `PublicAbstractMethodsHaveTypeChecker` rule is here: https://github.com/apache/spark/pull/12389/files

### CODE REVIEW

**Comment by lw-lin** [comment]
As per [Scalastyle - Custom rules](http://www.scalastyle.org/custom-rules.html), I installed the jar on my local machine, and verified that the added rule works.

Now I'm not sure how to pre-install the jar for every Spark build on Jenkins's machine -- should some build scripts be modified? @rxin could you give some hints? Thank you!

File: pom.xml [hunk 1]
```diff
@@ -2237,6 +2237,13 @@
           <inputEncoding>${project.build.sourceEncoding}</inputEncoding>
           <outputEncoding>${project.reporting.outputEncoding}</outputEncoding>
         </configuration>
+        <dependencies>
+          <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>scala_style_checker</artifactId>
```
Comment (srowen) [inline pom.xml]: Yikes, do we have to maintain and publish this as a module to use it? it may not be worth it

Comment (lw-lin) [inline pom.xml]: @srowen yeah I think you're right -- doing that is heavy and not worth it. :-)
I'm working on a light-wight manner to do this, after all all we need is to install the custom rules jar locally in the dev & build environment, other than publishing that into run-time environment or maven-central.
Thanks for the review!

**Comment by lw-lin** [comment]
hi @vanzin 
I'm not sure how to build and pre-install some jar for every Spark build on Jenkins's machine -- should some build scripts be modified? Could you give some hints when you have time? Thanks! :-)

(with some custom being built and pre-installed, we then can enable custom code style checks, please see http://www.scalastyle.org/custom-rules.html)

**Comment by rxin** [comment]
we can push this into scalastyle and then upgrade the version.

**Comment by vanzin** [comment]
@lw-lin I don't know either, but that sounds like too much work for style rules. I suggest following Reynold's advice and posting a PR to the scalastyle project; they're a little slow to update things, though.

**Comment by lw-lin** [comment]
I see; will follow @rxin 's advise. So I'm closing this.

@rxin @vanzin @srowen thank you all for the comments! :-)

**Pull request closed without merging**